src/python/pants/goal/stats_aggregator.py

```diff
@@ -149,6 +162,12 @@ def __call__(
                 f"{percentile_to_vals}"
             )
 
+        if self.output_file:
+            with open(self.output_file, "w") as fh:
+                fh.write(output_contents)
```

Comment:
As this doesn't happen in a any rule, I can't use https://www.pantsbuild.org/2.18/docs/writing-plugins/the-rules-api/file-system#workspacewrite_digest-save-to-disk, can I? Would this be an appropriate way to write into the build root directory?

Reply:
Potentially the `construct_callback` could request the `Workspace` and pass it in, although I don't know if holding on to that object outside the single rule invocation is a problem (i.e. if it's stashed in the callback for later).

Reply:
Thanks! I've experimented a bit, but it was a mess adding a workspace writing digest call as I have to construct a Digest outside of rule as well and to have it, notable amount of refactoring would be required which I'd like to avoid. It looks like `open` is used often in the repo to write things in the build root, so I assume this approach should be legit.

Reply:
yea, the thing to consider is if this may be called as part of a remote process invocation (i.e. on a remote build server rather than the local machine where `pants` was invoked.) In this case, I imagine that it being processed during the "finished" state, that it will be on the local machine only.. and as such, using `open` is legit. c.f. https://github.com/pantsbuild/pants/blob/e30da1ce11dfbe091e1b72772dc034d66d2b43e9/src/python/pants/core/goals/publish.py#L288-L290